src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1429,6 +1441,21 @@ protected MembersAndInitializers GetMembersAndInitializers()
             return membersAndInitializers!;
         }
 
+        [Conditional("DEBUG")]
+        internal void AssertMemberExposure(Symbol member)
+        {
+            var declared = _lazyDeclaredMembersAndInitializers;
+            Debug.Assert(declared != DeclaredMembersAndInitializers.UninitializedSentinel);
```

Comment:
The lack of a `Volatile.Read` caught my eye reading this method. There are a number of other places in the code where we access this member via `Volatile.Read`. I presume that's because we are accounting for the fact the current thread isn't guaranteed to hit one of the paths where we `CompareExchange` the value and hence could get a stale read here. 

The one call to this method is from the `Binder` and I'm unsure if that is on a path where we call through `BuildMembersAndInitializers`. If not I think we need a `Volatile.Read` here. Or am I missing another way that it's accounted for on this code path?

Reply:
There is only one call site for this function and it is not in ```Binder```. Regardless, it wouldn't hurt to add some volatile reads here to avoid any possibility of a false assert. I'll make this adjustment in the next PR. 

---
In reply to: [569945538](https://github.com/dotnet/roslyn/pull/50961#discussion_r569945538) [](ancestors = 569945538)